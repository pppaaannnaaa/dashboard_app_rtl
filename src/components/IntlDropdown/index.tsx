import React, { useContext } from 'react';
import ReactCountryFlag from 'react-country-flag';
import { Box, MenuItem, FormControl, OutlinedInput } from '@mui/material';
import Select, { SelectChangeEvent } from '@mui/material/Select';

// ** Internationalization Context
import { IntlContext } from '../../utility/context/Internationalization';
import { useRTL } from '../../utility/hooks';
import { MUIWrapperContext } from '../MUIWrapper';

// ** Vars
const langTitles: { [Key: string]: string } = {
  en: 'USA',
  ae: 'UAE',
  de: 'Germany',
  fr: 'France'
}

// ** lang with directions
export const names: { [Key: string]: string }[] = [
  { lang: "en", direction: "ltr" },
  { lang: "ae", direction: "rtl" },
  { lang: "de", direction: "ltr" },
  { lang: "fr", direction: "ltr" },
];

export const Themes: { [Key: string]: any } = {
  "en": {
    name: 'english',
    palette: { primary: { main: "#1976d2" }, secondary: { main: "#769fc7" } },
  },
  "ae": {
    name: 'arabic',
    palette: { primary: { main: "#665a6f" }, secondary: { main: "#9986a7" } },
  },
  "de": {
    name: 'german',
    palette: { primary: { main: "#303f9f" }, secondary: { main: "#656ea8" } },
  },
  "fr": {
    name: 'frence',
    palette: { primary: { main: "#d32f2f" }, secondary: { main: "#d16d6d" } },
  }
}


const MultipleSelectPlaceholder = () => {
  const intlContext = useContext(IntlContext);
  const [, setIsRtl] = useRTL();
  const { setLocale } = React.useContext(MUIWrapperContext);
  const { updateTheme } = React.useContext(MUIWrapperContext);

  // ** Function to switch Language
  const handleChange = (event: SelectChangeEvent) => {
    const { target: { value } } = event;

    intlContext.switchLanguage(value);
    updateTheme(Themes?.[value]);

    const _locale = names.find((name: { [Key: string]: string }) => name.lang === value);
    setLocale(_locale);

    // ** Update language Direction
    setIsRtl(!!(_locale?.direction === "rtl"));
  };

  return (
    <div>
      <FormControl sx={{ width: 200 }}>
        <Select
          displayEmpty
          onChange={handleChange}
          input={<OutlinedInput />}
          renderValue={() => (
            <Box display="flex" alignItems="center">
              <ReactCountryFlag countryCode={intlContext?.locale === 'en' ? 'us' : intlContext?.locale} svg />
              <Box mx={2}>{langTitles[intlContext?.locale]}</Box>
            </Box>
          )}
        >
          {
            names.map((name) => (
              <MenuItem key={name?.lang} value={name?.lang}>
                <ReactCountryFlag countryCode={name?.lang === 'en' ? 'us' : name?.lang} svg />
                <Box mx={2}>{langTitles[name?.lang]}</Box>
              </MenuItem>
            ))
          }
        </Select>
      </FormControl>
    </div>
  );
}

export default MultipleSelectPlaceholder;
