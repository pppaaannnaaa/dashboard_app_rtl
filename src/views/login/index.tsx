import React, { useState, useContext } from 'react';
import axios from 'axios';
import { toast } from 'react-toastify';
import { useDispatch } from 'react-redux';
import { useNavigate } from "react-router-dom";
import { FormattedMessage } from 'react-intl';
import { useForm, Controller, SubmitHandler } from "react-hook-form";
import { Avatar, Button, CssBaseline, TextField, Paper, Box, Grid, Typography, FormControl } from '@mui/material';
import LockOutlinedIcon from '@mui/icons-material/LockOutlined';

import { LOGIN_API, DASHBOARD_URL, EMAIL_REGEX } from '../../utility/constant';
import { handleLogin } from '../../redux/actions';
import { IntlContext } from '../../utility/context/Internationalization';

const IntlDropdown = React.lazy(() => import('../../components/IntlDropdown'));
const Loader = React.lazy(() => import('../../components/Loader'));
const ToggleMode = React.lazy(() => import('../../components/ToggleMode'));
const Signup = React.lazy(() => import('../../components/Signup'));

interface IFormInputs {
  email: string
  password: string
}

const Login = () => {
  const intlContext = useContext(IntlContext);
  const dispatch = useDispatch<any>();
  const navigate = useNavigate();

  const [loading, setLoading] = useState<boolean>(false);
  const [openSignup, setOpenSignup] = useState<boolean>(false);

  const { handleSubmit, control } = useForm<IFormInputs>({
    defaultValues: {
      email: "",
      password: ""
    },
  });

  const onSubmit: SubmitHandler<IFormInputs> = (data: IFormInputs) => {
    setLoading(true);
    axios
      .post(
        LOGIN_API,
        { email: data?.email, password: data?.password }
      )
      .then((res: any) => {
        setLoading(false);
        dispatch(handleLogin(res?.data));
        navigate(DASHBOARD_URL);
      })
      .catch((error: any) => {
        console.log(error);
        setLoading(false);
        const _error = error?.response?.data?.message || error?.message;
        toast(_error, { type: "error" })
      });
  }

  const validateEmail = (value: any) => {
    const regex = EMAIL_REGEX[intlContext.locale];
    const isValid = regex.test(value);

    if (!isValid) {
      return `Please enter valid email with .${intlContext.locale}`
    }
  };

  return (
    <Box>
      <Loader enable={loading} />

      <Grid container component="main" sx={{ height: '100vh' }}>
        <CssBaseline />

        <Grid
          item
          xs={false}
          sm={4}
          md={7}
          sx={{
            backgroundImage: 'url(https://source.unsplash.com/random?wallpapers)',
            backgroundRepeat: 'no-repeat',
            backgroundColor: (t) =>
              t.palette.mode === 'light' ? t.palette.grey[50] : t.palette.grey[900],
            backgroundSize: 'cover',
            backgroundPosition: 'center',
          }}
        />

        <Grid item xs={12} sm={8} md={5} component={Paper} elevation={6} square>
          <Box
            sx={{
              my: 6,
              mx: 4,
              display: 'flex',
              flexDirection: 'column',
              alignItems: 'center',
            }}
          >
            <Box mb={12} display="flex" alignItems="center" justifyContent="space-between" width="100%">
              <IntlDropdown />
              <Box display="flex" alignItems="center">
                <Box ml={2}><ToggleMode /></Box>
              </Box>
            </Box>

            <Avatar sx={{ m: 1, bgcolor: 'secondary.main' }}>
              <LockOutlinedIcon />
            </Avatar>

            <Typography component="h1" variant="h5">
              <FormattedMessage id="sign_in" />
            </Typography>

            <Box
              component="form"
              onSubmit={handleSubmit(onSubmit)}
              sx={{ mt: 1 }}
            >
              <FormControl fullWidth>
                <Controller
                  name="email"
                  control={control}
                  rules={{ required: true, validate: validateEmail }}
                  render={({
                    field: { onChange, value },
                    fieldState: { error }
                  }) => (
                    <TextField
                      margin="normal"
                      autoFocus
                      autoComplete="email"
                      label={<FormattedMessage id="email" />}
                      value={value}
                      helperText={error ? error.message : null}
                      error={!!error}
                      onChange={onChange}
                    />
                  )}
                />
              </FormControl>

              <FormControl fullWidth>
                <Controller
                  name="password"
                  control={control}
                  rules={{ required: true }}
                  render={({
                    field: { onChange, value },
                    fieldState: { error }
                  }) => (
                    <TextField
                      type="password"
                      margin="normal"
                      label={<FormattedMessage id="password" />}
                      value={value}
                      helperText={error ? error.message : null}
                      error={!!error}
                      onChange={onChange}
                    />
                  )}
                />
              </FormControl>

              <Button
                type="submit"
                fullWidth
                size='large'
                variant="contained"
                sx={{ mt: 3, mb: 2 }}
              >
                {<FormattedMessage id="sign_in" />}
              </Button>

              <Box display="flex" justifyContent="flex-end">
                <Button onClick={() => setOpenSignup(true)} style={{ textTransform: "inherit" }}>
                  <FormattedMessage id="no_ac_sign_up" />
                </Button>
              </Box>

            </Box>
          </Box>
        </Grid>
      </Grid>

      {openSignup ? <Signup open={openSignup} setOpen={setOpenSignup} /> : null}
    </Box>
  );
}

export default Login;
