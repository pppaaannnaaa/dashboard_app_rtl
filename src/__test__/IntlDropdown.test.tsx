import React from 'react';
import { cleanup } from '@testing-library/react';
import { renderWithReactIntl } from './login.test';
import IntlDropdown from '../components/IntlDropdown';

afterEach(cleanup);

describe('IntlDropdown', () => {
  it('should match snapshot', async () => {
    const { asFragment } = renderWithReactIntl(<IntlDropdown />);
    expect(asFragment()).toMatchSnapshot();
  });
});
