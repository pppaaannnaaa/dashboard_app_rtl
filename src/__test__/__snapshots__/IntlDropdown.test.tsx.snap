// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntlDropdown should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="MuiFormControl-root css-1orw215-MuiFormControl-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <div
          aria-controls=":r2:"
          aria-expanded="false"
          aria-haspopup="listbox"
          class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          role="combobox"
          tabindex="0"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <div
              class="MuiBox-root css-1at62qq"
            />
          </div>
        </div>
        <input
          aria-hidden="true"
          aria-invalid="false"
          class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          tabindex="-1"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Login should match snapshot 1`] = `<DocumentFragment />`;
